src/parser/control_parser.rs

```diff
@@ -19,20 +21,25 @@ pub fn parse_control_statement(lexer: &mut ParseSession) -> Result<Statement, Di
     }
 }
 
-fn parse_if_statement(lexer: &mut ParseSession) -> Result<Statement, Diagnostic> {
+fn parse_if_statement(lexer: &mut ParseSession) -> Statement {
     let start = lexer.range().start;
     lexer.advance(); //If
     let mut conditional_blocks = vec![];
 
     while lexer.last_token == KeywordElseIf || lexer.last_token == KeywordIf {
-        let condition = parse_expression(lexer);
-        lexer.expect(KeywordThen)?;
+        let condition = parse_primary_expression(lexer);
+        expect_token!(
```

Comment:
I don't think we need this: 
You can use parse in region or a similar method (there's one for statements) with a closing condition KeywordThen

Reply:
the parse_in_region is not good for this situation because it will try to recover until the next safe end-keywords eating everything in between:
```
IF condition
    call1();
    call2();
 END_IF
```
 
 the missing THEN would try to recover until the END_IF and ignore call1() and call2()
 
 take a look how the parse_while_statement(...) treats the situation:
```
WHILE condition DO
     ...
END_WHILE
```
 
 it enforces an expression after the while (and reports manually if it cannot find any) and consume_or_report the DO-keyword therefore surviving a missing DO-keyword.
 only then it enters a end-while region (although maybe all of this could already be wrapped into the end-while-region)